keps/sig-scheduling/3280-guarantee-pdb-when-preemption-happens/README.md

```diff
@@ -0,0 +1,962 @@
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+<!--
+This section is incredibly important for producing high-quality, user-focused
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+This design proposal suggests adding a field `PreemptionPolicy` in the `PodDisruptionBudget` (PDB) API to indicate 
+whether or not make `PodDisruptionBudget` guaranteed during the scheduler preemption process, in order to meet the 
+needs of users who do not want PDB violations to occur during scheduler preemption due to some reasons such as 
+high availability of services(https://github.com/kubernetes/kubernetes/issues/91492#issuecomment-1029484252).
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+PodDisruptionBudget (PDB) is used to limit the number of concurrent disruptions that your application experiences, 
+allowing for high availability. Users can set the field '.spec.maxUnavailable' or '.spec.minAvailable' to declare 
+the current minimum availability or maximum unavailability to be maintained after eviction. 
+
+However, there is currently an issue where the kube-scheduler does not strictly guarantee PDBs during the preemption 
+process. The scheduler supports PDBs when preempting pods, but the adherence to PDBs is best effort. The scheduler 
```

Comment:
Does the scheduler do this by inspecting PDBs and replicating the PDB controller logic, or by preempting pods via the /evictions endpoint?

Reply:
replicating logic

Reply:
I thinks scheduler not just replicates the PDB controller logic, it inspects the PDB status and preempt the pods via the /pod/delete api.

https://github.com/kubernetes/kubernetes/blob/724497cda467b72bf56271637ec5edb00c19cedf/pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go#L289-L335

Do you suggest to remove it via /evictions endpoint? @liggitt

Reply:
We should not use the /evictions endpoint, as that will again respect PDB, which we need to violate at least for system priority.

Maybe there is chance to refactor some of the PDB logic so that it can be reused by scheduler. We should aim to do that.

Reply:
> We should not use the /evictions endpoint, as that will again respect PDB, which we need to violate at least for system priority.

scheduler should use the evictions endpoint to preempt a pod while strictly honoring PDB

Reply:
Ah, right. So we will have two deletion paths.

We need to make sure the scheduler cache works correctly if /evictions rejects the deletion.

Reply:
Yes. When the user declares the need for strict adherence to PDB, in order to avoid concurrent issues maybe arising from the controller, the final decision will still need to be made by the /evictions endpoint. However, if the user does not make such a declaration, pods will continue to be directly removed through the use of the remove method, as per the current implementation.

Reply:
I think it might be useful to call api eviction also in the best effort case and if that fails fallback to the deletion.

It seems that the main problem of https://github.com/kubernetes/kubernetes/issues/91492 is the slow disruption controller reaction to the `status.disruptionsAllowed` which can race with the scheduler and cause incorrect disruptions.

So in the best effort case:
- if `disruptionAllowed > 0` the eviction will work and the api will decrement the PDB status, and the scheduler should watch and pick up this change almost instantly
- if `disruptionAllowed == 0` then we will fallback to deletion and the race should not occur since we are already at minimum

Also, if we start using the eviction API in all cases, we can sort the pods by ready status, and prefer the unready pods. The eviction would acknowledge a recently added field [unhealthyPodEvictionPolicy](https://kubernetes.io/docs/tasks/run-application/configure-pdb/#unhealthy-pod-eviction-policy) in PDBs which is relevant when `disruptionAllowed == 0`, and do least ammount of damage.

Reply:
> if disruptionAllowed == 0 then we will fallback to deletion and the race should not occur since we are already at minimum.

Do you mean when the value of disruptionAllowed is 0, the /eviction endpoint will return a response that rejects this option, and the scheduler will need to fallback to deletion as an alternative. Right?

> The eviction would acknowledge a recently added field [unhealthyPodEvictionPolicy](https://kubernetes.io/docs/tasks/run-application/configure-pdb/#unhealthy-pod-eviction-policy) in PDBs which is relevant when disruptionAllowed == 0, and do least ammount of damage.

Thanks, we have been working to minimize the impact of preemption on healthy running services, and unhealthyPodEvictionPolicy are a good option, though they may need to be considered for integration and use after they are released in beta.

Reply:
> Do you mean when the value of disruptionAllowed is 0, the /eviction endpoint will return a response that rejects this option, and the scheduler will need to fallback to deletion as an alternative. Right?

yes

Reply:
@denkensk can you please mention this under the BestEffort policy?

Reply:
Done @atiratree